### PR TITLE
8328168: Epsilon: Premature OOM when allocating object larger than uncommitted heap size

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

When an object size meets the following three conditions, the allocation will fail unexpectedly.

1. `object_size` > `uncommitted_size`
2. `object_size` > `committed_but_unused_size`
3. `object_size` < `unused_size` (`unused_size` equals `uncommitted_size + committed_but_unused_size`)

Please read the newly added test of this patch for the concrete case.

Thanks for taking the time to review.

Best Regrads,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328168](https://bugs.openjdk.org/browse/JDK-8328168): Epsilon: Premature OOM when allocating object larger than uncommitted heap size (**Bug** - P2)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18304/head:pull/18304` \
`$ git checkout pull/18304`

Update a local copy of the PR: \
`$ git checkout pull/18304` \
`$ git pull https://git.openjdk.org/jdk.git pull/18304/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18304`

View PR using the GUI difftool: \
`$ git pr show -t 18304`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18304.diff">https://git.openjdk.org/jdk/pull/18304.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18304#issuecomment-1997552959)

### CODE REVIEW

File: test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ */
+
+package gc.epsilon;
+
+/**
+ * @test TestEnoughUnusedSpace
+ * @requires vm.gc.Epsilon
+ * @summary Epsilon should allocates object successfully if it has enough space.
+ * @run main/othervm -XX:InitialHeapSize=64M -Xmx128M -XX:+UnlockExperimentalVMOptions
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java:31]: Do we really need `-XX:InitialHeapSize`, or `-Xms64M` would work as well?

File: test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *                   -XX:+UseEpsilonGC gc.epsilon.TestEnoughUnusedSpace
+ */
+
+public class TestEnoughUnusedSpace {
+    public static void main(String[] args) {
+        // Create an array about 90M. It should be created successfully
+        // instead of throwing OOME, because 90M is smaller than 128M.
+        byte[] arr = new byte[90 * 1024 * 1024];
+        // Just prevent the compiler optimization.
+        System.out.println(arr.length);
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java:41]: This is also amenable for quite easy compiler optimization. See how other Epsilon tests do it: they sink the object into `static volatile Object sink;`.

File: src/hotspot/share/gc/epsilon/epsilonHeap.cpp [hunk 3]
```diff
@@ -125,17 +125,18 @@ HeapWord* EpsilonHeap::allocate_work(size_t size, bool verbose) {
 
       // Expand and loop back if space is available
       size_t size_in_bytes = size * HeapWordSize;
-      size_t space_left = max_capacity() - capacity();
+      size_t uncommitted_space = max_capacity() - capacity();
+      size_t unused_space = max_capacity() - used();
```
Comment (shipilev) [inline src/hotspot/share/gc/epsilon/epsilonHeap.cpp:129]: Please also: 

```
assert(unused_space <= uncommitted_space, 
       "Unused (" SIZE_FORMAT ") <= uncommitted (" SIZE_FORMAT ")",
       unused_space, uncommitted_space);
```

**Review by shipilev — APPROVED** [review]
Oh. So the critical change is here:

```
} else if (size_in_bytes < unused_space) {
  // No space to expand in bulk, and this allocation is still possible,
  // take all the remaining space:
```

Before this fix, we might have had some unused, but already expanded space. But we would miss this, since we only checked how much there was left to expand to. Which means we still have space to afford the allocations, we instead we just throw a premature OOM. Whoops.

I suggest we rename the issue to `Epsilon: Premature OOM when allocating object larger than uncommitted heap size`.

**Review by shipilev — CHANGES_REQUESTED** [review]
(meant to request changes)

File: test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java (hunk 1 above)
Comment (lgxbslgx) [reply test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java]: Changed to `-Xms64M`.

File: test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java (hunk 2 above)
Comment (lgxbslgx) [reply test/hotspot/jtreg/gc/epsilon/TestEnoughUnusedSpace.java]: > This is also amenable for quite easy compiler optimization.

You are right, my mistake.The length of this array can be statically determined.

>  See how other Epsilon tests do it: they sink the object into `static volatile Object sink;`.

Fixed.

File: src/hotspot/share/gc/epsilon/epsilonHeap.cpp (hunk 3 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/epsilon/epsilonHeap.cpp:129]: I think it should be `assert(unused_space >= uncommitted_space, ...)`.

**Comment by lgxbslgx** [comment]
> ⚠️ @lgxbslgx This pull request contains merges that bring in commits not present in the target repository. Since this is not a "merge style" pull request, these changes will be squashed when this pull request in integrated. If this is your intention, then please ignore this message. If you want to preserve the commit structure, you must change the title of this pull request to `Merge <project>:<branch>` where `<project>` is the name of another project in the [OpenJDK organization](https://github.com/openjdk) (for example `Merge jdk:master`).

This problem can be solved after https://github.com/openjdk/jdk/pull/18303 integrating.

File: src/hotspot/share/gc/epsilon/epsilonHeap.cpp (hunk 3 above)
Comment (shipilev) [reply src/hotspot/share/gc/epsilon/epsilonHeap.cpp:129]: Right!

**Review by shipilev — APPROVED** [review]
This looks fine, thanks.

**Comment by lgxbslgx** [comment]
@shipilev Thanks for your review. Waiting for another review.

**Review by tschatzl — APPROVED** [review]

**Comment by lgxbslgx** [comment]
@shipilev @tschatzl Thanks for the review.

/integrate

**Pull request closed without merging**